wlroots/renderer.py

```diff
@@ -36,13 +36,26 @@ def clear(self, color) -> None:
         color_ptr = ffi.new("float[4]", color)
         lib.wlr_renderer_clear(self._ptr, color_ptr)
 
+    def render_texture(
+        self, texture: Texture, projection: Matrix, x: int, y: int, alpha: float
+    ) -> bool:
+        """Renders the requested texture"""
+        ret = lib.wlr_render_texture(
+            self._ptr, texture._ptr, projection._ptr, x, y, alpha
+        )
+        if not ret:
+            # TODO: get a better exception type
+            raise Exception("Bad render")
+        return ret
+
     def render_texture_with_matrix(
         self, texture: Texture, matrix: Matrix, alpha: float
-    ) -> None:
+    ) -> bool:
```

Comment:
Is there a reason to need to return this? If it fails, then the exception is thrown.

Reply:
Hmm good point. I think returning it and not using an exception more closely mirrors wlroots. With the exception, the user has to right code to catch it whether they want to continue or not, just in case it does error. Without the exception, they would just have to check the return value if they would want to do something responding to the error. Just thinking how it would actually be used -- likely in a rendering loop on an output's frame event or similar. If a texture isn't rendered, you'd likely want the rest of the loop to continue, so this would require a try... pass block as is, or nothing extra if it doesn't raise the exception. I think you're right, the return value is unnecessary here, but I guess it makes sense to use that rather than an exception. What do you think?

Reply:
Actually I changed my mind, I think your original is the best way, using an Exception. I wouldn't _want_ the renderer to continue if it wasn't working for some reason. Presumably if one rendering call fails, they all will, and it's worth more investigation. That shouldn't be happening in normal use anyway.

Reply:
Right, my thinking on this was that using an exception is a bit more of a Pythonic way to handle error conditions, in Raymond's [P vs NP way](https://www.youtube.com/watch?v=wf-BqAjZb8M&t=766s). If it is the case that this returns `false` relatively frequently as a part of normal operation, it might make sense to re-evaluate the exception, but having the exception for a failure IMO is the more expected behavior, at least for a Python API, and can help to eliminate subtle issues because in some part of the code we forgot to check the return value.

Reply:
Yeah sounds good.